chore: add transformers

KFlash opened the pull request:
This implements an transformer API and a visitor to transform the `RootFile`.

It will transform and get rid of Kataw's own types, but can be extended and new transformers can be added.

TODO:
-------


- [x]  Traverse and update all CST nodes on changes

- [x]  Handle lexical scoping

- [x]  Allow swapping of keywords

- [x]  Get rid of all Kataw types

- [ ]  Performance

KFlash commented:
@aladdin-add The goal is to replace current walker with the one in his PR. Can you find a way to test the transformers?

KFlash commented:
@aladdin-add Removing Kataw types now works. Can you test this?  I'm going to finish up the transformer now and focus on it's performance

aladdin-add commented on src/transform/visitor.ts:
```diff
@@ -0,0 +1,1259 @@
+} from './core';
+
+export function visitEachChild(transform: Transform, node: any, visitor: (node: SyntaxKind) => SyntaxKind): any {
+  const kind = node.kind;
```
have you tested it? the type node: SyntaxKind seems not correct.

KFlash replied:
I'm working on testing and improve it locale. There are bugs that I fixed already, but not pushed

KFlash replied:
The SyntaxKind is correct, you need to do a type cast to get it correct. Se my update

KFlash commented:
@aladdin-add Can you add this to the test runner?

aladdin-add commented:
> Can you add this to the test runner?

not sure what you mean by "add this to the test runner". can you outline what you want to test, and what the API looks like?

aladdin-add commented:
maybe off-topic: we cannot test all the things in the snapshots - we will have to introduce something like mocha/jest/node-tap to writing some unit tests.

KFlash commented:
@aladdin-add You remove the current walker and use the new on instead in the test runner - visitor.ts. Then just walk !

KFlash commented:
@aladdin-add then add back mocha

KFlash commented:
@aladdin-add The visitor should be okay. I fixed the types and corrected a bunch of bugs. Can you replace the old walker?

And I think we merge this as is and do follow up PR for lexicals and other transformers

KFlash commented:
@aladdin-add I added a 'removeKatawTypes' as public API. See the readme. It need to be tested in the test runner

I'm merging this now
